connector/signaltometricsconnector/connector.go

```diff
@@ -5,46 +5,253 @@ package signaltometricsconnector // import "github.com/open-telemetry/openteleme
+							dp := dps.At(l)
+							filteredDPAttrs, ok := md.FilterAttributes(dp.Attributes())
+							if !ok {
+								continue
+							}
+							multiError = errors.Join(multiError, aggregate(dp, filteredDPAttrs))
+						}
+					case pmetric.MetricTypeEmpty:
+						multiError = errors.Join(multiError, fmt.Errorf("metric %q: invalid metric type: %v", metric.Name(), metric.Type()))
+					}
+				}
+			}
+		}
+	}
+	aggregator.Finalize(sm.dpMetricDefs)
+	return sm.processNext(ctx, processedMetrics, multiError)
 }
 
-func (sm *signalToMetrics) ConsumeLogs(context.Context, plog.Logs) error {
-	return nil
+func (sm *signalToMetrics) ConsumeLogs(ctx context.Context, logs plog.Logs) error {
+	if len(sm.logMetricDefs) == 0 {
+		return nil
+	}
+
+	var multiError error
+	processedMetrics := pmetric.NewMetrics()
+	processedMetrics.ResourceMetrics().EnsureCapacity(logs.ResourceLogs().Len())
+	aggregator := aggregator.NewAggregator[ottllog.TransformContext](processedMetrics)
+	for i := 0; i < logs.ResourceLogs().Len(); i++ {
+		resourceLog := logs.ResourceLogs().At(i)
+		resourceAttrs := resourceLog.Resource().Attributes()
+		for j := 0; j < resourceLog.ScopeLogs().Len(); j++ {
+			scopeLog := resourceLog.ScopeLogs().At(j)
+			for k := 0; k < scopeLog.LogRecords().Len(); k++ {
+				log := scopeLog.LogRecords().At(k)
+				logAttrs := log.Attributes()
+				for _, md := range sm.logMetricDefs {
+					filteredLogAttrs, ok := md.FilterAttributes(logAttrs)
+					if !ok {
+						continue
+					}
+
+					// The transform context is created from original attributes so that the
+					// OTTL expressions are also applied on the original attributes.
+					tCtx := ottllog.NewTransformContext(log, scopeLog.Scope(), resourceLog.Resource(), scopeLog, resourceLog)
+					if md.Conditions != nil {
+						match, err := md.Conditions.Eval(ctx, tCtx)
+						if err != nil {
+							multiError = errors.Join(multiError, fmt.Errorf("failed to evaluate conditions, skipping: %w", err))
+							continue
+						}
+						if !match {
+							sm.logger.Debug("condition not matched, skipping", zap.String("name", md.Key.Name))
+							continue
+						}
+					}
+
+					filteredResAttrs := md.FilterResourceAttributes(resourceAttrs)
+					multiError = errors.Join(multiError, aggregator.Aggregate(ctx, tCtx, md, filteredResAttrs, filteredLogAttrs, 1))
+				}
+			}
+		}
+	}
+	aggregator.Finalize(sm.logMetricDefs)
+	return sm.processNext(ctx, processedMetrics, multiError)
+}
+
+// processNext is a helper method for all the Consume* methods to do error handling,
+// logging, and sending the processed metrics to the next consumer in the pipeline.
+func (sm *signalToMetrics) processNext(ctx context.Context, m pmetric.Metrics, err error) error {
+	if err != nil {
+		dpCount := m.DataPointCount()
+		if dpCount == 0 {
+			// No signals were consumed so return an error
+			return fmt.Errorf("failed to consume signal: %w", err)
+		}
+		// At least some signals were partially consumed, so log the error
```

Comment:
Does it make sense to emit partial metrics when errors have occurred? 🤔 I'm a bit worried about the consistency and the correctness of this approach.

In any case, can we document this logic in the README?

Reply:
> Does it make sense to emit partial metrics when errors have occurred?

The issue here is that the metrics are already partially consumed and if we error out then they might be retried. Since, there is no good way to convey partial failure that I am aware of so the only alternative would be to somehow revert the state of the store...

I will add a readme but happy to discuss if you have some other ideas/solution in mind.

Reply:
Updated the readme in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36852/commits/d768c5d49cc40d8bcaf12725c6b1778e1d267c76.

> I'm a bit worried about the consistency and the correctness of this approach

Regarding consistency and correctness, I am not sure I understand the concern. Ideally, a payload with partial fail should always produce the same result: partial metric with a warning log.

Reply:
My concern is that partial metrics can be misleading if the users do not know they are partial. The warning log is helpful but not all users will be aware of it. In addition it will not be easy to "corelate" partial metrics with these warning logs if you are trying to interpret some "odd" metrics that you see in your back-end.

Maybe we could make this partial emission configurable and let users decide if they want partial metrics or not? (this would mean that on error the Consume* does not aggregate and forward anything)

Reply:
Hmm, I don't want to give a configuration option tbh. Thinking again on your concern, I think it makes a lot of sense to not have partial metrics behaviour by default. Let me go back to the code and see if I can fix that.

Reply:
Done in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36852/commits/5d4c73cc410b538586638446309ebfeb2c48828c, let me know if it looks better now.

Reply:
That's also fine. We can always extend this behavior in the feature if we see the need for it.

Thank's for addressing this!